### PR TITLE
[FLINK-34954][core] Kryo Input bug fix

### PR DESCRIPTION
**Pull request by qinghui-xu**
Handle edge case of zero length serialized bytes correctly.

## What is the purpose of the change

Bug fix in kryo (NoFetching)Input implementation to handle properly zero length serialized bytes, eg the serialization of a protobuf message with default values.

## Brief change log

  - Fix while loop for `NoFetchingInput#read(byte[], int, int)` and `NoFetchingInput#require(int)`

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: yes
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by qinghui-xu** [comment]
Hello team,
Could I get a review for this, please?

**Review by dannycranmer — APPROVED** [review]
LGTM, thanks @qinghui-xu

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/NoFetchingInput.java [hunk 1]
```diff
@@ -73,17 +73,14 @@ protected int require(int required) throws KryoException {
         position = 0;
         int bytesRead = 0;
         int count;
-        while (true) {
+        while (bytesRead < required) {
             count = fill(buffer, bytesRead, required - bytesRead);
```
Comment (davidradl) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/NoFetchingInput.java:77]: @dannycranmer I saw this linked to in slack - I wondered whether we could have junits for the required case and higher than required? Maybe as a follow on issue / pr ? I could code if you were willing to merge for me?

**Comment by dmvk** [comment]
Why is this not covered with a regression test? That feels rather scary in such a critical part of the stack.

**Pull request merged**